### PR TITLE
Fix #3295: Introduce UI for Developer Options Menu

### LINKED ISSUE #3295: Introduce UI for Developer Options Menu
Introduce UI for Developer Options Menu.

**Mock link:** https://xd.adobe.com/view/e8aa4198-3940-47f9-514a-f41cc54457f6-9e9b/screen/116991fa-d0eb-42f9-9db2-b4456d797a48/

**Target PR date:** 2021-06-14
**Target completion date:** 2021-06-18

### PR DESCRIPTION
**Pull request by yashraj-01**
## Explanation

Fixes #3295 : Introduced UI for Developer Options Menu

**Mock link:** https://xd.adobe.com/view/e8aa4198-3940-47f9-514a-f41cc54457f6-9e9b/screen/116991fa-d0eb-42f9-9db2-b4456d797a48/

**Screenshots**

- UI
  

**All test results are after adding** `AccessibilityTestRule`

- Espresso test passing -- `NavigationDrawerActivityDebugTest`
Screenshot from 2021-06-30 23-42-41

- Robolectric test passing -- `NavigationDrawerActivityDebugTest`
Screenshot from 2021-06-30 23-45-33

- Espresso test passing -- `DeveloperOptionsActivityTest`
Screenshot from 2021-06-30 23-51-10

- Robolectric test passing -- `DeveloperOptionsActivityTest`
Screenshot from 2021-06-30 23-51-42

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

File: app/src/sharedTest/java/org/oppia/android/app/devoptions/DeveloperOptionsActivityTest.kt
Comment (Sarthak2601) [inline app/src/sharedTest/java/org/oppia/android/app/devoptions/DeveloperOptionsActivityTest.kt:144]: I think we can add a few more tests to check the hints and solutions' switch status. These can be similar to the ones we do for wifi switch in administrator controls. Additionally in the future (when the corresponding activities are made), we'll need to add tests related to opening of appropriate activity on option selection in this test file as well.

**Review by Sarthak2601 — CHANGES_REQUESTED** [review]
Great @yashraj-01! LGTM for the most part. Left a comment for adding a few more tests. 
Also, some tests seem to be failing -- PTAL, thanks.

File: app/src/sharedTest/java/org/oppia/android/app/devoptions/DeveloperOptionsActivityTest.kt
Comment (yashraj-01) [reply app/src/sharedTest/java/org/oppia/android/app/devoptions/DeveloperOptionsActivityTest.kt:144]: Okay. I'll add these tests and also check for failing tests.

Comment (yashraj-01) [reply app/src/sharedTest/java/org/oppia/android/app/devoptions/DeveloperOptionsActivityTest.kt:144]: I have added the test for the switch. PTAL. Thanks.

Comment (Sarthak2601) [reply app/src/sharedTest/java/org/oppia/android/app/devoptions/DeveloperOptionsActivityTest.kt:144]: Aah, I think there was one more test involving config change.

Comment (yashraj-01) [reply app/src/sharedTest/java/org/oppia/android/app/devoptions/DeveloperOptionsActivityTest.kt:144]: Yes. But that test is feasible only after implementing the logic for showing hints and solution. We can add that test when we implement this or we can add the test now and mark it as **Ignore**.

Comment (Sarthak2601) [reply app/src/sharedTest/java/org/oppia/android/app/devoptions/DeveloperOptionsActivityTest.kt:144]: Oh okay, maybe we can put it in place at the time of logic implementation.

**Review by Sarthak2601 — APPROVED** [review]
LGTM.

**Review by fsharpasharp — APPROVED** [review]
Approving for BUILD.bazel change

**Review by rt4914 — CHANGES_REQUESTED** [review]
Will review this tomorrow.

**Comment by vinitamurthi** [comment]
Hi @yashraj-01, Is there a reason I was asked to review this PR? I see Rajat and Sarthak are already doing it. It seems like none of the files are files I codeown. You can instead just add me as CC when my review isn't required. Unassigning myself now!

**Comment by yashraj-01** [comment]
> Hi @yashraj-01, Is there a reason I was asked to review this PR? I see Rajat and Sarthak are already doing it. It seems like none of the files are files I codeown. You can instead just add me as CC when my review isn't required. Unassigning myself now!

Oh okay. Thanks for letting me know about this.

File: app/src/main/java/org/oppia/android/app/devoptions/DeveloperOptionsActivity.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/android/app/devoptions/DeveloperOptionsActivity.kt]: Name these as per convention mentioned in this issue: #1587

Comment (rt4914) [reply app/src/main/java/org/oppia/android/app/devoptions/DeveloperOptionsActivity.kt]: Also possibly mention in alphabetical order.

File: app/src/main/java/org/oppia/android/app/devoptions/devoptionsitemviewmodel/DeveloperOptionsItemViewModel.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/android/app/devoptions/devoptionsitemviewmodel/DeveloperOptionsItemViewModel.kt]: Add `.` at the end.

**Review by rt4914 — CHANGES_REQUESTED** [review]
@yashraj-01  Took overview pass. Implementation looks good. Will take a detailed pass once previous two PRs are merged.

**Comment by BenHenning** [comment]
See https://github.com/oppia/oppia-android/pull/3314#issuecomment-869992586. I'm not sure what I should be reviewing here--please clarify @yashraj-01 / @Sarthak2601.

**Comment by yashraj-01** [comment]
> See [#3314 (comment)](https://github.com/oppia/oppia-android/pull/3314#issuecomment-869992586). I'm not sure what I should be reviewing here--please clarify @yashraj-01 / @Sarthak2601.

Apologies for the inconvenience. Will update the reviewers and assignee sections.

File: app/src/main/java/org/oppia/android/app/devoptions/DeveloperOptionsActivity.kt
Comment (yashraj-01) [reply app/src/main/java/org/oppia/android/app/devoptions/DeveloperOptionsActivity.kt]: I was refactoring all the PRs and it came to my notice that these are unnecessary. I have removed them now.

File: app/src/main/java/org/oppia/android/app/devoptions/devoptionsitemviewmodel/DeveloperOptionsItemViewModel.kt
Comment (yashraj-01) [reply app/src/main/java/org/oppia/android/app/devoptions/devoptionsitemviewmodel/DeveloperOptionsItemViewModel.kt]: Done.

**Review by rt4914 — CHANGES_REQUESTED** [review]
@yashraj-01 Make sure that all Espresso tests support `AccessibilityTestRule`.

You will need to add it in your test file 
```
@get:Rule
  val accessibilityTestRule = AccessibilityTestRule()
```

Once you add this you will notice that one test fails because the `Hints/Solution` switch is less than 48x48 dp. Once this is fixed in UI, these tests will pass. Add the screenshot of passing tests in description.

**Comment by yashraj-01** [comment]
> @yashraj-01 Make sure that all Espresso tests support `AccessibilityTestRule`.
> 
> You will need to add it in your test file
> 
> ```
> @get:Rule
>   val accessibilityTestRule = AccessibilityTestRule()
> ```
> 
> Once you add this you will notice that one test fails because the `Hints/Solution` switch is less than 48x48 dp. Once this is fixed in UI, these tests will pass. Add the screenshot of passing tests in description.

Done. Also added new test results in PR description.

**Review by rt4914 — APPROVED** [review]
LGTM, thanks.

**Comment by rt4914** [comment]
@BenHenning This PR needs your approval for Bazel file which was earlier approved by @fsharpasharp

File: app/src/sharedTest/java/org/oppia/android/app/devoptions/DeveloperOptionsActivityTest.kt
Comment (anandwana001) [inline app/src/sharedTest/java/org/oppia/android/app/devoptions/DeveloperOptionsActivityTest.kt:224]: can we write a test case to check click operation on this switch?

**Review by anandwana001 — CHANGES_REQUESTED** [review]
Apart from adding UI, do we have click operation implemented in this PR on these recycler view items?

If yes, please write the test case for them too.

File: app/src/sharedTest/java/org/oppia/android/app/devoptions/DeveloperOptionsActivityTest.kt
Comment (yashraj-01) [reply app/src/sharedTest/java/org/oppia/android/app/devoptions/DeveloperOptionsActivityTest.kt:224]: Actually, the click operations are not implemented on any of the views in this PR. So, I think we cannot write tests to check click operation yet. WDYT?

Comment (anandwana001) [reply app/src/sharedTest/java/org/oppia/android/app/devoptions/DeveloperOptionsActivityTest.kt:224]: ok, could you add here which issue will be tracking the user interaction?

Comment (yashraj-01) [reply app/src/sharedTest/java/org/oppia/android/app/devoptions/DeveloperOptionsActivityTest.kt:224]: Sure. will do that.

**Review by vinitamurthi — APPROVED** [review]

**Review by anandwana001 — CHANGES_REQUESTED** [review]
Assign me back once you add the issue tracking the interaction thing.

File: app/src/sharedTest/java/org/oppia/android/app/devoptions/DeveloperOptionsActivityTest.kt
Comment (yashraj-01) [reply app/src/sharedTest/java/org/oppia/android/app/devoptions/DeveloperOptionsActivityTest.kt:224]: Added TODO. PTAL. The issue is #3397.

**Comment by yashraj-01** [comment]
> Assign me back once you add the issue tracking the interaction thing.

I have added the issue tracking the click operation. PTAL. Thanks.

File: app/src/main/java/org/oppia/android/app/devoptions/devoptionsitemviewmodel/DeveloperOptionsModifyLessonProgressViewModel.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/devoptions/devoptionsitemviewmodel/DeveloperOptionsModifyLessonProgressViewModel.kt]: For this & related view models, this seems too non-specific. Please update the class KDocs to more specifically describe what the view model represents (since presumably its presence in the recycler view results in something changing for the UI displayed to the user).

File: app/src/main/java/org/oppia/android/app/devoptions/DeveloperOptionsViewModel.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/devoptions/DeveloperOptionsViewModel.kt:21]: This is technically still part of the public API of this view model, so it ought to have a KDoc.

File: app/src/main/java/org/oppia/android/app/devoptions/DeveloperOptionsViewModel.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/devoptions/DeveloperOptionsViewModel.kt]: What does the view model represent with regards to the UI? The KDoc should be more specific.

File: app/src/main/java/org/oppia/android/app/devoptions/DeveloperOptionsViewModel.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/devoptions/DeveloperOptionsViewModel.kt]: Mutability isn't needed here:

```suggestion
    return listOf(
      DeveloperOptionsModifyLessonProgressViewModel(),
      DeveloperOptionsViewLogsViewModel(),
      DeveloperOptionsOverrideAppBehaviorsViewModel()
    )
```

**Review by BenHenning — COMMENTED** [review]
Thanks @yashraj-01. I looked at the affected BUILD file & just the files being added to the BUILD file.

**Review by anandwana001 — APPROVED** [review]
LGTM

File: app/src/main/java/org/oppia/android/app/devoptions/DeveloperOptionsViewModel.kt
Comment (yashraj-01) [reply app/src/main/java/org/oppia/android/app/devoptions/DeveloperOptionsViewModel.kt]: Done.

File: app/src/main/java/org/oppia/android/app/devoptions/DeveloperOptionsViewModel.kt
Comment (yashraj-01) [reply app/src/main/java/org/oppia/android/app/devoptions/DeveloperOptionsViewModel.kt]: KDoc update. PTAL.

File: app/src/main/java/org/oppia/android/app/devoptions/DeveloperOptionsViewModel.kt
Comment (yashraj-01) [reply app/src/main/java/org/oppia/android/app/devoptions/DeveloperOptionsViewModel.kt:21]: KDoc added. PTAL.

File: app/src/main/java/org/oppia/android/app/devoptions/devoptionsitemviewmodel/DeveloperOptionsModifyLessonProgressViewModel.kt
Comment (yashraj-01) [reply app/src/main/java/org/oppia/android/app/devoptions/devoptionsitemviewmodel/DeveloperOptionsModifyLessonProgressViewModel.kt]: KDoc updated. PTAL.

**Review by BenHenning — APPROVED** [review]
Thanks @yashraj-01. LGTM for codeowners (BUILD files).

**Pull request merged**
